Add appointment command

khoongwk opened the pull request:
Major update to include classes of Storage, Logic, Ui, and Model for the Appointment class.

Current status: `addappt` command works in the GUI. Entries are stored into `appointments.json`.

However, there is an issue which was uncovered: When a duplicate appointment is added into 
Homerce, the existing `appointments.json` is cleared.

khoongwk committed:
Merge branch 'master' of https://github.com/AY2021S1-CS2103T-W13-3/tp

* 'master' of https://github.com/AY2021S1-CS2103T-W13-3/tp:
  Fix minor format mistake
  Fix minor formatting mistake
  Update User Guide
  Fix mistake when merging
  Update user guide for revenue

khoongwk committed:
Merge branch 'master' of https://github.com/AY2021S1-CS2103T-W13-3/tp

khoongwk committed:
Add JUnit Tests for Time

khoongwk committed:
Add JUnit Test Skeleton for Appointment

Skeleton Test class for Appointment. Fill in when logic is done.

khoongwk committed:
Add CLI Syntax for Appointment

khoongwk committed:
Refactor Time.java to TimeOfDay.java

Renamed Time class for clarity.

khoongwk committed:
Add More OOP for Appointment's Status

khoongwk committed:
Add Ui Components for Appointment

To display appointments in the GUI, let's add AppointmentCard and
AppointmentListpanel, which handles the display of appointments
onto JavaFX.

khoongwk committed:
Add Appointment Manager Classes

To store Appointments into SuperSalon, let's create Manager classes
to store appointments into a UniqueList.

khoongwk committed:
Add Appointment into ModelManager

To manipulate Appointments stored in the Model, ModelManager must contain
an AppointmentManager. Let's add methods into ModelManager to handle
Appointments.

khoongwk committed:
Add AddAppointmentCommand, Edit ModelManager

Added command to add appointment and edited Model classes to include
Appointment Manager.

khoongwk committed:
Refactor Appointment, Add Parser for AddAppointmentCommand

AddAppointmentCommandParser will create a Command containing an Appointment object.
The command will be executed to store into the model and storage.

However, it is not possible to create Client and Service objects within ParserUtil as attributes
for Appointment, as a search is needed to be performed within Model to match the input Phone
and input ServiceCode to the right Client and Service. Thus, let's refactor Appointment to
include Phone and ServiceCode as attributes, while making Client and Service Optional.

khoongwk committed:
Update Logic, AddressBookParser

Added methods to include Appointments

khoongwk committed:
Add UniqueServiceList

Service list is currently an implementation of a generic Unique List.
To find a service by service code, we need to overload the contains
method to accept a service code. Let's create a unique service class
to contain this method.

khoongwk committed:
Complete Model and Command for AddAppointment

khoongwk committed:
Add Storage Classes for Appointment

Homerce (our new name) needs to have a model for Appointments in JSON format.

Let's add in Json-adapted classes for Appointment.

khoongwk committed:
Complete Add Appointment Command

Storage components were implemented and other bugs fixed.

Current issue remains after light testing: Adding of duplicate appointment will result
in clearing of the entire appointments.json storage file.

Test classes which use the hasClient method to test for null have been commented out
due to ambiguity with overloaded method parameters.

galvinleow commented on src/main/java/seedu/address/model/manager/ServiceManager.java:
```diff
@@ -8,25 +8,26 @@
 public class ServiceManager implements ReadOnlyServiceManager {
 
-    private final UniqueList<Service> services;
+    private final UniqueServiceList services;
 
     public ServiceManager() {
-        this.services = new UniqueList<>();
+        this.services = new UniqueServiceList();
     }
```
I think Hans made a generic list?

khoongwk replied:
Yup he did, but I needed to extend it with more methods. I need a contains method with parameter ServiceCode to check if the service code belongs to an existing service. I can't just insert that as a generic method to Unique List, so I created UniqueServiceList to extend UniqueList<Service>.

hanstirtaputra reviewed (approved):
LGTM

khoongwk committed:
Added List Appointment Command

khoongwk committed:
Add Find Appointment Command

Predicate classes made for the find appointment command.

Missing finding by month feature, will add in the future.

khoongwk committed:
Fix Prefix checking for Find Appointment Command

Prefix checking in FindCommandParser was not checking for invalid
prefixes and not checking for empty parameters. Let's add in checks
using the argMultimap

khoongwk committed:
Add Delete Appointment Command

khoongwk committed:
Add Clear Appointment Command

galvinleow commented on src/main/java/seedu/address/logic/parser/appointment/FindAppointmentCommandParser.java:
```diff
@@ -0,0 +1,87 @@
+            && !arePrefixesPresent(argMultimap, PREFIX_SERVICE_SERVICE_CODE))
+            || !argMultimap.getPreamble().isEmpty()) {
+            throw new ParseException(String.format(MESSAGE_INVALID_COMMAND_FORMAT,
+                AddAppointmentCommand.MESSAGE_USAGE));
```
FindAppointmentCommand

khoongwk replied:
Fixed, thanks!

khoongwk committed:
Add Done, Undone, Edit Commands for Appointment

RuiFengg reviewed (approved):
LGTM
